### ISSUE TITLE
data: Synthetic data generation for desk

### ISSUE THREAD

**Issue by tikikun** [menloresearch/space-thinker#6]
* EB-Manipulation extends [VLMBench](https://github.com/eric-ai-lab/VLMbench) to evaluate MLLM-based embodied agents in low-level object manipulation. 

* How to do it? 
   * Action space discretization, which divides the position components **(x,y,z) into 100 bins** and the orientation components **(roll,pitch,yaw) into 120 bins**, enabling valid actions to take forms like [x,y,z,roll,pitch,yaw,gripper] = [57,61,20,10,60,25,1]
   * Additional information like YOLO detection boxes with index markers and object pose estimation for indexed objects, reducing the need for precise 3D location.

| <img width="880" alt="Image" src="https://github.com/user-attachments/assets/b5e3782b-6cf4-4dee-a52d-c2d954028bea" /> | 
|:--:| 
| The vision-driven agent pipeline used in [EMBODIEDBENCH](https://github.com/EmbodiedBench/EmbodiedBench) |

**Comment by tuanlda78202** [comment]
VLMBench will be used to generate a desk dataset, but it lacks support for 7D coordinates (EmbodiedBench converts these to 6 axes via a custom PyTorch dataset). We must implement a discretization pipeline to convert continuous values to discrete coordinates within [0, 100]

![Image](https://github.com/user-attachments/assets/e8aa9acc-8008-4601-a724-23ad2aa96c93)

We also want to simplify desk interactions. Rather than dealing with a 100x100 desk, we'll generate data using a 10x10 grid. Once we reach a local cell, we'll subdivide it into another 10x10 grid, and so on.

![Image](https://github.com/user-attachments/assets/5e523186-9c00-41e4-8878-4dd858943b13)

![Image](https://github.com/user-attachments/assets/c713f3cc-7958-43df-b668-78cf50c63657)

**Comment by tuanlda78202** [comment]
The following are visualizations of the simulation

| <img width="600" alt="Image" src="https://github.com/user-attachments/assets/1aa27aa0-2b03-4ae2-baf3-efe5d4576c5d" /> | 
|:--:| 
| Stack cubes only do two objects stack |

| <img width="600" alt="Image" src="https://github.com/user-attachments/assets/c42a147a-e9d8-4d04-9e3e-1315bfcd882a" /> | 
|:--:| 
| Place into shape sorter |

**Comment by new5558** [comment]
## Video Visualization of the synthesized training dataset (From VLMBench)
**Pick Cube**
![Image](https://github.com/user-attachments/assets/664e5399-5f53-431c-a8f9-3ecda7d172a9)

**Place Into Shape Sorter**
![Image](https://github.com/user-attachments/assets/a4413bb8-6e0a-4780-b8d0-109b2671e9de)

**Stack Cubes**
![Image](https://github.com/user-attachments/assets/1ca9c114-3f0e-4259-91bd-c88c2c529695)

**Wipe Table**
![Image](https://github.com/user-attachments/assets/1d7f333f-6520-4c6f-a508-ebbbac258420)

**Comment by tuanlda78202** [comment]
# Visualization of benchmarking on EmbodiedBench (Claude 3.7)

* EB-Manipulation includes four task categories:
    * Pick & Place Objects
    * Stack Objects
    * Shape Sorter Placement
    * Table Wiping

* Each of the remaining 4 subsets comprises 48 tasks evenly distributed across four
categories, with 12 tasks per category.

## Pick up the right object and place it into the right container (2)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/a96cc20a-352d-4fde-a04e-f06615115978" /> | 
|:--:| 
| Success |

## Pick up the right object and place it into the rear container (5)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/05ff9f6e-ec4c-4f62-8a88-1d77c796ac6c" /> | 
|:--:| 
| Fail |

## Stack the right triangular prism on top of the right moon (14)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/ce22ea24-24e1-450d-b9ee-101915b4fda8" /> | 
|:--:| 
| Success |

## Stack the left star on top of the right triangular prism (15)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/bffb57c5-2266-45ad-9d6e-4c3583b95ff2" /> | 
|:--:| 
| Fail |

## Put the left star into the shape sorter (28)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/8f51d2c7-c18d-4d29-9ead-322679df236a" /> | 
|:--:| 
| Fail |

## Put the rear star into the shape sorter (32)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/93175964-ea58-45eb-939c-105a358322d6" /> | 
|:--:| 
| Success |

## Wipe the right area (38)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/ccc7c033-9428-481d-b9f0-30ba7694b3a8" /> | 
|:--:| 
| Success |

## Wipe the left area (40)

| <img width="800" alt="Image" src="https://github.com/user-attachments/assets/bc28bfb1-cd17-4f67-86f9-6da12f3b2846" /> | 
|:--:| 
| Fail |

**Comment by tuanlda78202** [comment]
# Data Format

* Format
    * [Yaw, Pitch, Roll] in range [-180, 180]

```bash
Robot Arm:
- p1 -> [x, y] * n_sub_task
- p1_z -> [z] * n_sub_task
- p1_rotation -> [Yaw, Pitch, Roll] * n_sub_task

Target:
- p0 -> [x, y] * n_sub_task ---- use for training + inference Clip6D model
- p0_z -> [z] * n_sub_task ---- unused at all
- p0_rotation -> [Yaw, Pitch, Roll] * n_sub_task ---- unused at all
```

* Example 
```bash
'p0': array([[69, 27],
       [69, 27]], dtype=int16), 
'p0_z': array([[0.08159541], # 8 cm above
       [0.08159541]]), 
'p0_rotation': array([[89.99932427, 13.76822344, 90.01462889],
       [89.99932427, 13.76822344, 90.01462889]]), 
'p1': array([[ 72,  25],
       [121,  39]], dtype=int16), 
'p1_z': array([[0.09708737],
       [0.17065995]]), 
'p1_rotation': array([[ 9.90175635e+01, -1.90868402e-02, -1.79990773e+02],
       [-1.37722815e+02,  7.02052689e-03,  1.79997260e+02]])} 
```

**Comment by tuanlda78202** [comment]
# Discrete <-> Continuous Action

```python
import numpy as np
from scipy.spatial.transform import Rotation

SCENE_BOUNDS = np.array([-0.3, -0.5, 0.6, 0.7, 0.5, 1.6]) # Defines the 3D workspace boundaries [x_min, y_min, z_min, x_max, y_max, z_max]
ROTATION_RESOLUTION = 3 # Angle step size in degrees for discretizing rotations
VOXEL_SIZE = 100

def discrete_to_quaternion(discrete_action):
    pos_indicies = np.array(discrete_action[:3])
    rot_and_grip_indicies = np.array(discrete_action[3:6])
    is_gripper_open = discrete_action[6]

    res = (SCENE_BOUNDS[3:] - SCENE_BOUNDS[:3]) / VOXEL_SIZE
    attention_coordinate = SCENE_BOUNDS[:3] + res * pos_indicies + res / 2
    
    # discrete_euler to quaternion
    # This shifts the range from [0, 360] back to [-180, 180] degrees, which is the standard range for Euler angles that the Rotation class expects
    euluer = (rot_and_grip_indicies * ROTATION_RESOLUTION) - 180
    quat = Rotation.from_euler('xyz', euluer, degrees=True).as_quat()
    
    continuous_action = np.concatenate([
        attention_coordinate,
        quat,
        [is_gripper_open]
    ])
    return continuous_action

discrete_action = [28, 80, 17, 
                   0, 60, 94, 
                   1]
print(discrete_to_quaternion(discrete_action))

def quaternion_to_discrete(continuous_action):
    """
    Convert continuous action to discrete action format.
    
    Args:
        continuous_action: numpy array containing:
            - position [x,y,z] (first 3 elements)
            - quaternion [qx,qy,qz,qw] (next 4 elements) 
            - gripper state (last element)
    
    Returns:
        discrete_action: list containing:
            - position indices [x,y,z] (first 3 elements)
            - rotation indices [rx,ry,rz] (next 3 elements)
            - gripper state (last element)
    """
    position = continuous_action[:3]
    quaternion = continuous_action[3:7]
    is_gripper_open = continuous_action[7]

    # Convert position to voxel indices
    res = (SCENE_BOUNDS[3:] - SCENE_BOUNDS[:3]) / VOXEL_SIZE
    trans_indices = np.floor((position - SCENE_BOUNDS[:3]) / res).astype(np.int32)
    
    # Ensure indices are within bounds
    trans_indices = np.clip(trans_indices, 0, VOXEL_SIZE - 1)

    # Convert quaternion to euler angles and then to discrete indices
    euler = Rotation.from_quat(quaternion).as_euler('xyz', degrees=True)
    # Map from [-180, 180] to [0, 360/ROTATION_RESOLUTION]
    rot_indices = np.floor((euler + 180) / ROTATION_RESOLUTION).astype(np.int32)
    # Ensure rotation indices are within bounds
    rot_indices = np.clip(rot_indices, 0, 360//ROTATION_RESOLUTION - 1)

    # Combine into final discrete action
    discrete_action = np.concatenate([
        trans_indices,    # position indices [x,y,z]
        rot_indices,      # rotation indices [rx,ry,rz]
        [is_gripper_open] # gripper state
    ]).tolist()

    return discrete_action

cons_8d_action = np.array([-0.015, 0.305, 0.775, 
                           -0.62932, -0.77715, 4.7586e-17, 3.8535e-17, 
                           1.0])
print(quaternion_to_discrete(cons_8d_action))

def discretize(action, image_res=500):
    """
    Convert action from [x, y, z, Yaw, Pitch, Roll] format to discrete action format.
    
    Args:
        action: list containing [x, y, z, Yaw, Pitch, Roll]
            - x, y: pixel coordinates (depends on image resolution)
            - z: float in range [0,1]
            - Yaw, Pitch, Roll: angles in degrees, range [-180, 180]
        image_res: image resolution (assuming square image)
    
    Returns:
        discrete_action: list containing 7 elements:
            - position indices [x,y,z] (first 3 elements)
            - rotation indices [rx,ry,rz] (next 3 elements)
            - gripper state (last element, default to 1)
    """
    x, y, z, yaw, pitch, roll = action
    
    # First convert x,y from image coordinates (pixels) to scene coordinates
    x_scene = SCENE_BOUNDS[0] + (x / image_res) * (SCENE_BOUNDS[3] - SCENE_BOUNDS[0])
    y_scene = SCENE_BOUNDS[1] + (y / image_res) * (SCENE_BOUNDS[4] - SCENE_BOUNDS[1])
    
    # Convert z from [0,1] to actual z-bounds
    z_scene = SCENE_BOUNDS[2] + z * (SCENE_BOUNDS[5] - SCENE_BOUNDS[2])
    
    # Convert scene coordinates to voxel indices
    res = (SCENE_BOUNDS[3:] - SCENE_BOUNDS[:3]) / VOXEL_SIZE
    x_idx = int(np.floor((x_scene - SCENE_BOUNDS[0]) / res[0]))
    y_idx = int(np.floor((y_scene - SCENE_BOUNDS[1]) / res[1]))
    z_idx = int(np.floor((z_scene - SCENE_BOUNDS[2]) / res[2]))
    
    # Clip position indices to valid range
    x_idx = np.clip(x_idx, 0, VOXEL_SIZE - 1)
    y_idx = np.clip(y_idx, 0, VOXEL_SIZE - 1)
    z_idx = np.clip(z_idx, 0, VOXEL_SIZE - 1)
    
    # Convert rotation angles to indices
    rot_indices = np.floor((np.array([yaw, pitch, roll]) + 180) / ROTATION_RESOLUTION).astype(np.int32)
    rot_indices = np.clip(rot_indices, 0, 360//ROTATION_RESOLUTION - 1)
    
    # Combine into final discrete action (adding gripper state as 1)
    discrete_action = np.concatenate([
        [x_idx, y_idx, z_idx],  # position indices
        rot_indices,            # rotation indices
        [1]                     # gripper state (default to 1/open)
    ]).tolist()
    
    return discrete_action

# Example usage:
vlmbench_action = [69, 27, 0.08159541, 89.99932427, 13.76822344, 90.01462889]
discrete_action = discretize(vlmbench_action, image_res=500)
print("Discrete action:", discrete_action)
```

**Comment by new5558** [comment]
## Where X,Y,Z of the each objects come from:

**Input:**
- Depth Cameras `[500, 500] X N cemeras` -> Estimate point cloud `[500, 500,3] X N cemeras, 3 = 3D coordinates`,
- RGB Mask Cemeras (Segmentaion of each object) `[500,500] X N cemeras`
There are N cameras: `['front', 'left_shoulder', 'right_shoulder', 'wrist']`
- Mapper which mask id is which object ex.`{90: 'first container', 93: 'second container', 97: 'first cube', 99: 'second cube'} mask_id_to_real_name`

**Process:**
1. Use RGB mask to seperate each object's point cloud data
2. Average all point cloud of the objects in the same cemera
3. Average output from 2) of all cameras
4. Convert point cloud to voxel format using point_to_voxel_index
    - Capped max at 100

```python
SCENE_BOUNDS = np.array([-0.3, -0.5, 0.6, 0.7, 0.5, 1.6])
VOXEL_SIZE = 100

# From https://github.com/stepjam/RLBench/blob/master/rlbench/backend/utils.py
def point_to_voxel_index(
        point: np.ndarray):
    print(point, 'point')
    bb_mins = np.array(SCENE_BOUNDS[0:3])[None]
    bb_maxs = np.array(SCENE_BOUNDS[3:])[None]
    dims_m_one = np.array([VOXEL_SIZE] * 3)[None] - 1
    bb_ranges = bb_maxs - bb_mins
    res = bb_ranges / (np.array([VOXEL_SIZE] * 3) + 1e-12)
    voxel_indicy = np.minimum(
        np.floor((point - bb_mins) / (res + 1e-12)).astype(
            np.int32), dims_m_one)

    return voxel_indicy.reshape(point.shape)
 ```

NOTE: https://en.wikipedia.org/wiki/Voxel

Example of each object if Converted "Object 1" -> "Real Name"
```
[('second cube', [40, 24, 18]), ('second container', [68, 41, 17]), ('first cube', [31, 57, 18]), ('first container', [43, 72, 17])] avg_obj_coord
```

**Comment by tikikun** [comment]
## Global and local context

<img width="1061" alt="Image" src="https://github.com/user-attachments/assets/74b61ab5-64be-4b70-8908-07dc876b625b" />

**Comment by bachvudinh** [comment]
Try to synthetic the data using [VLM Bench](https://github.com/eric-ai-lab/VLMbench/blob/main/tools/dataset_generator_NLP.py). the script under the hood using Pyrep- a Python library that acts as a bridge between your Python and the CoppeliaSim robotics simulator and AmSolver-Python framework built on top of PyRep. It's specifically designed for generating synthetic data for robotic manipulation tasks.

Unable to save scene and configuration file due to an error on the Prep side. Issue persists despite deploying on different machines and attempting to build Docker with Ubuntu:22.04 as base:
```

FROM nvidia/cuda:12.4.0-runtime-ubuntu20.04

```

We end up with the same bug:

![Image](https://github.com/user-attachments/assets/d2eb048e-4d33-4c76-b4d1-e24c17760056)

**Comment by bachvudinh** [comment]
We plan to replicate their data pipeline in Genesis. I'm currently familiarizing myself with Genesis while studying the VLMBench data pipeline to understand their data synthesis approach. Our initial goal is to spawn objects in collision-free positions and implement pick-and-place tasks with the Panda robot. The expected data output will be in JSON format:
```
{
"Desk_size": 
"Object": [{"description": position}, ...],
"instruction: ,
"solution": [],
}
```
Example:
{
"Desk_size": 100x100x30
"Object": [{"red star": [35, 15, 17]}, {"green cube": [57, 18, 18]}, {"red container ": [76, 69, 17]}, {"purple container": [49, 37, 17]}, {"orange cylinder": [54, 81, 18]} ],
"instruction:  "Pick up the star and place it into the red container",
"solution": [[35, 15, 30, 0, 60, 0, 1], [35, 15, 17, 0, 60, 0, 1], [35, 15, 17, 0, 60, 0, 0], [35, 15, 30, 0, 60, 0, 0], [76, 69, 30, 0, 60, 0, 0], [76, 69, 17, 0, 60, 0, 0], [76, 69, 17, 0, 60, 0, 1]],
}

**Comment by bachvudinh** [comment]
## Detailed information of some objects:

1. Dining Table
table_height=0.752
URDF file: [diningTable.zip](https://github.com/user-attachments/files/19203413/diningTable.zip)
  ![Image](https://github.com/user-attachments/assets/c65b8206-0f04-4ecb-96c1-9315c137b319)

2. Container
URDF file: [container.urdf.zip](https://github.com/user-attachments/files/19203425/container.urdf.zip)
3. Moon object.
URDF file: [moon.zip](https://github.com/user-attachments/files/19203429/moon.zip)
